canvas: Apply large size limitations

tharkum opened the pull request:
To prevent any potential crash/OOM issues with "canvas" element
from "rogue" applications let's apply large size limitations for context
canvas2d's draw target to Servo (similar approach in Firefox/Chromium -
they limits width and height to 32767/65535 pixels).

Fixes: #36155, #34117, #30164, #24710

--
- [x] ./mach build -d does not report any errors
- [x] ./mach test-tidy does not report any errors
- [x] There are tests for these changes
tests/wpt/tests/html/canvas/element/canvas-host/2d.canvas.host.size.large.html
tests/wpt/tests/html/canvas/offscreen/canvas-host/2d.canvas.host.size.large.html
tests/wpt/tests/html/canvas/offscreen/canvas-host/2d.canvas.host.size.large.worker.js

tharkum commented:
cc @xiaochengh, @sagudev < Please review.

xiaochengh commented on components/script/dom/htmlcanvaselement.rs:
```diff
@@ -371,14 +372,26 @@ impl HTMLCanvasElement {
+            None => false,
+        };
+
+        is_paintable || is_valid_canvas_size(self.get_size().to_u64())
```
When do we check `is_valid_canvas_size()`?

tharkum replied:
This is tricky thing.

We should assume canvas is paintable ("createImageBitmap/toBlob/toDataURL" will return transparent black image as snapshot with canvas size) if associated context also is paintable otherwise if canvas element has not context need to check against supported limits.

However there are some nitpick for WebGL context - event if canvas size is 0x0 the WebGL context will be successfully created and paintable!, so need to add additional check here

> HTMLCanvasElement.width and .height values less than 1 are treated as 1. A 0x0 canvas will yield a 1x1 drawingBufferWidth/Height.
https://registry.khronos.org/webgl/specs/latest/1.0/#2.2

```
pub(crate) fn is_paintable(&self) -> bool {
   ...
   (paintable && !size.is_empty()) || is_supported_canvas_size(size.to_u64())
}
```

xiaochengh replied:
If this is specific to WebGL, could you move it to WebGL context's impl of `is_paintable`?

tharkum replied:
Done

tharkum commented on components/shared/canvas/canvas.rs:
```diff
@@ -79,7 +79,7 @@ pub struct CanvasId(pub u64);
     Canvas2d(Canvas2dMsg, CanvasId),
     FromScript(FromScriptMsg, CanvasId),
-    Recreate(Option<Size2D<u64>>, CanvasId),
+    Recreate(Option<Size2D<u64>>, bool, CanvasId),
```
Like this? `Recreate(Option<Size2D<u64>>, bool /* reset_state */, CanvasId),`

tharkum commented on components/script/canvas_state.rs:
```diff
@@ -178,7 +206,8 @@ impl CanvasState {
         script_to_constellation_chan
             .send(ScriptToConstellationMessage::CreateCanvasPaintThread(
-                size, sender,
+                Size2D::zero(),
```
The zero here because of lazy creation of draw target for "2d" context.
At the start we will don't allocate memory for draw target  on canvas paint thread side.

PaintRenderingContext2D for paint worklets do the similar:
https://github.com/servo/servo/blob/main/components/script/dom/paintrenderingcontext2d.rs#L47

Look like it is only one place there are we call this message, so I will  remove "size" parameter at all.

tharkum commented on components/script/canvas_state.rs:
```diff
@@ -242,15 +336,27 @@ impl CanvasState {
         self.ipc_renderer
-            .send(CanvasMsg::Recreate(Some(size), self.get_canvas_id()))
+            .send(CanvasMsg::Recreate(
+                Some(Size2D::zero()),
```
Originally "Recreate" message was for resizing of  draw target, but later "Reset" functionality was added into it, so

```
Recreate(Some(X)) -> resizing drawing buffer + reset canvas states
Recreate(None) -> reallocate drawing buffer with previous size + reset canvas states
```

tharkum commented on components/script/canvas_state.rs:
```diff
@@ -221,14 +251,78 @@ impl CanvasState {
+    }
+
     pub(crate) fn send_canvas_2d_msg(&self, msg: Canvas2dMsg) {
+        if !self.is_paintable() {
```
Script thread (CanvasState) -> Canvas thread (CanvasData)

If our draw target is not paintable, our rendering context also is not paintable, so we able to skip any Canvas2D messages (except Recreate message) until user will resize canvas back to supported size.
At the same time any canvas state for non paintable context could be fetch from CanvasState on script thread side. 

line_width/set_line_width: https://github.com/servo/servo/blob/main/components/script/canvas_state.rs#L1231

xiaochengh replied:
Ack, thanks!

tharkum commented:
@xiaochengh < I decided to simplify PR and remove the "lazy allocation of canvas draw target" functionality
and prepare another PR for it.
So now this PR has only "is_paintable" validation for HTMLCanvasElement.

xiaochengh commented on components/script/canvas_state.rs:
```diff
@@ -60,6 +60,39 @@ use crate::dom::paintworkletglobalscope::PaintWorkletGlobalScope;
+        }
+    }
+
+    fn get_size(&self) -> Size2D<u64> {
```
nit: rename to `get_paintable_size` since it's not a trivial getter

tharkum replied:
Done

xiaochengh commented on components/script/dom/offscreencanvas.rs:
```diff
@@ -121,8 +120,15 @@ impl OffscreenCanvas {
+            _ => false,
+        };
+
+        (paintable && !size.is_empty()) || is_supported_canvas_size(size)
```
Could you explain why OffScreenCanvas needs to check more than `paintable`?

tharkum replied:
You compared with outdated HTMLCanvasElement.rs (in the latest version they have similar check)
https://github.com/servo/servo/pull/36445/files#diff-7483a0f0ae501ab9da1432c28321c22f39647521072498f773903f360b026ce3R388

xiaochengh replied:
I'm still not following why some canvas types need extra `is_supported_canvas_size` checks and some do not.

In an early message you mentioned that WebGL needs this check as a fixup, in which case I believe it should be moved to WebGL related types.

If it's general, then could you elaborate the difference between the paintability of a DrawTarget and that of a canvas (be it HTMLCanvas or offscreen)?

tharkum replied:
HTMLCanvasElement "paintable" (top level)
   - ContextCanvas2D DrawTarget "paintable" (low level)
   - WebGL/WebGPU DrawingBuffer "paintable" (low level)
 

There are some places in code there we could/must check possibility to snapshot image data OR draw from other canvas and current proposal was to up checking of "paintability" to top level instead of waiting until it will be done at context level (but it will be more CPU resourceble because we can do validation twice, but could help on some edge cases).

NOTE:
"Paintable" on low level (per context) will be validated against draw target size limitations + context lost status (for GPU accelerated rendering - not yet supported on 100%)

xiaochengh reviewed (commented):
Thanks for splitting it up and making it easier to review!

sagudev commented:
> Canvas context is paintable if:
>
>    "2d": draw target is paintable (depends from canvas size)
>    "webgl": always true (no context lost tracking)
>    "webgpu": always true

Hm, do we really need `paintable` concept? Looking at this PR give me idea to simply move canvas size handling to internal get_image_data of contexts: https://github.com/sagudev/servo/commit/1f346416f0b2a964adb68ba8773cb474b13a1e41 that builds on top of https://github.com/servo/servo/pull/36119 (while still not fixing max size, but that should be only mater of additional checks in get_image_data of 2d canvas context). WDYT?

tharkum commented:
> > Canvas context is paintable if:
> > "2d": draw target is paintable (depends from canvas size)
> > "webgl": always true (no context lost tracking)
> > "webgpu": always true
> 
> Hm, do we really need `paintable` concept? Looking at this PR give me idea to simply move canvas size handling to internal get_image_data of contexts: [sagudev@1f34641](https://github.com/sagudev/servo/commit/1f346416f0b2a964adb68ba8773cb474b13a1e41) that builds on top of #36119 (while still not fixing max size, but that should be only mater of additional checks in get_image_data of 2d canvas context). WDYT?


Your change with snapshot concept is a bit large, so need time to check it well, but at first view "paintable" concept is more top-level compare with "snapshot".

Instead of doing "paintable" validation inside each context method implementation we will do it inside HTMLCanvasElement. 

"Paintable" concept will allow to do:
1. Preliminary check before call "get_image_data" from  "createImageBitmap/toBlob/toDataURL" methods
    and to return transparent black image ("snapshot::cleared") if canvas is not paintable.
2. Don't execute CanvasContext2D operations on non paintable draw target
3. Unify with canvas_data_source() - WR image key
https://github.com/servo/servo/pull/36445/files#diff-1175ac14bef74692069d7e4df307be02255c0c71f3776c2578db4e0dcda3ff95R108

Anyway I will check your PR and return back with more feedback.

sagudev commented:
> transparent black image ("snapshot::cleared") if canvas is not paintable

The problem here is that snapshot::cleared will do allocation, so in case of too big canvas we would still get OOM.

tharkum commented:
> > transparent black image ("snapshot::cleared") if canvas is not paintable
> 
> The problem here is that snapshot::cleared will do allocation, so in case of too big canvas we would still get OOM.

https://github.com/servo/servo/pull/36119/files#diff-7483a0f0ae501ab9da1432c28321c22f39647521072498f773903f360b026ce3R382

How about to replace it to this "is_paintable" check (included validation of max canvas size) to avoid OOM?
`if !self.paintable() ~~~~ ! (context_paintable || is_supported_canvas_size(size))`

sagudev commented:
> > > transparent black image ("snapshot::cleared") if canvas is not paintable
> > 
> > 
> > The problem here is that snapshot::cleared will do allocation, so in case of too big canvas we would still get OOM.
> 
> https://github.com/servo/servo/pull/36119/files#diff-7483a0f0ae501ab9da1432c28321c22f39647521072498f773903f360b026ce3R382
> 
> How about to replace it to this "is_paintable" check (included validation of max canvas size) to avoid OOM? `if !self.paintable() ~~~~ ! (context_paintable || is_supported_canvas_size(size))`

That, would work, but that's included in https://github.com/sagudev/servo/commit/1f346416f0b2a964adb68ba8773cb474b13a1e41 (which I pushed just now, but already linked before). Actually rather than putting spotlight on my PR I just wanted to put it on https://github.com/sagudev/servo/commit/1f346416f0b2a964adb68ba8773cb474b13a1e41, which does similar as paintable but for each context in get_image_data. I think this is important for two reasons:
- context have context specific requirements
- makes get_image_data on specific context without HTMLCanvasElement actually usable in edge cases.

So I am questioning reasoning for this:

> Instead of doing "paintable" validation inside each context method implementation we will do it inside HTMLCanvasElement.

Is there any reason why we cannot do this in each context and I am missing it?

tharkum commented:
> > > > transparent black image ("snapshot::cleared") if canvas is not paintable
> > > 
> > > 
> > > The problem here is that snapshot::cleared will do allocation, so in case of too big canvas we would still get OOM.
> > 
> > 
> > https://github.com/servo/servo/pull/36119/files#diff-7483a0f0ae501ab9da1432c28321c22f39647521072498f773903f360b026ce3R382
> > How about to replace it to this "is_paintable" check (included validation of max canvas size) to avoid OOM? `if !self.paintable() ~~~~ ! (context_paintable || is_supported_canvas_size(size))`
> 
> That, would work, but that's included in [sagudev@1f34641](https://github.com/sagudev/servo/commit/1f346416f0b2a964adb68ba8773cb474b13a1e41) (which I pushed just now, but already linked before). Actually rather than putting spotlight on my PR I just wanted to put it on [sagudev@1f34641](https://github.com/sagudev/servo/commit/1f346416f0b2a964adb68ba8773cb474b13a1e41), which does similar as paintable but for each context in get_image_data. I think this is important for two reasons:
> 
> * context have context specific requirements
> * makes get_image_data on specific context without HTMLCanvasElement actually usable in edge cases.
> 
> So I am questioning reasoning for this:
> 
> > Instead of doing "paintable" validation inside each context method implementation we will do it inside HTMLCanvasElement.
> 
> Is there any reason why we cannot do this in each context and I am missing it?

@sagudev < Sorry for long reply...

I viewed your "snapshot" PR and "yes" after your unification of "get_image_data" my proposed "paintable" concept
 for HTMLCanvasElement could be omitted, because we have only one useful consumer (get image data)
of this change on top-level and it is more convenient to check restrictions on context level instead
of HTML canvas element.

There are one possible (but not mandatory) consumer of it is "CanvasState::draw_image_internal" for HTMLCanvasElement/OffscreenCanvas, but it could also be done per context level right now (with software image drawing).

So I can propose to cut-off "HTMLCanvasElement::is_paintable" change from PR and preserve only "CanvasState::is_paintable" for DrawTarget (to validate canvas size).

sagudev commented:
> So I can propose to cut-off "HTMLCanvasElement::is_paintable" change from PR and preserve only "CanvasState::is_paintable" for DrawTarget (to validate canvas size).

Sounds good to me.

BTW I created test site that allow to set custom canvas size on different context, to see how other browsers respond to edge cases (and to see what they return on `toDataUrl`), but I haven't got the time to test any browser https://sagudev.github.io/briefcase/canvas-sizes.html (it does not work on servo tho).

tharkum commented:
> > So I can propose to cut-off "HTMLCanvasElement::is_paintable" change from PR and preserve only "CanvasState::is_paintable" for DrawTarget (to validate canvas size).
> 
> Sounds good to me.
> 
> BTW I created test site that allow to set custom canvas size on different context, to see how other browsers respond to edge cases (and to see what they return on `toDataUrl`), but I haven't got the time to test any browser https://sagudev.github.io/briefcase/canvas-sizes.html (it does not work on servo tho).

There are some existed application for testing canvas size:
https://jhildenbiddle.github.io/canvas-size/#/

tharkum committed:
canvas: Apply upper size limitation to context output bitmap

To prevent any potential crash/OOM issues with <canvas> element
from "rogue" applications let's apply upper size limitations
for rendering context's output bitmap (draw target) to Servo
(similar approach in Firefox/Chromium -
they limits width and height to 32767/65535 pixels).

Bug: #36155

Signed-off-by: Andrei Volykhin <andrei.volykhin@gmail.com>

xiaochengh reviewed (approved):
LGTM.

Thanks for further splitting the PR and making it nice and clean!

mrobinson commented on components/script/canvas_state.rs:
```diff
@@ -1892,6 +1951,19 @@ impl CanvasState {
 }
 
+pub(crate) fn is_supported_canvas_size(size: Size2D<u64>) -> bool {
+    if size.is_empty() {
+        return false;
+    }
+    if size.width > MAX_CANVAS_SIZE || size.height > MAX_CANVAS_SIZE {
+        return false;
+    }
+    if size.area() > MAX_CANVAS_AREA {
+        return false;
+    }
+    true
```
I would just combine this one:
```suggestion
    !size.is_empty() &&
        size <= Size2D::new(MAX_CANVAS_SIZE, MAX_CANVAS_SIZE) &&
        size.area() < MAX_CANVAS_AREA
```

tharkum replied:
And `size <= Size2D::new(MAX_CANVAS_SIZE, MAX_CANVAS_SIZE)` is not supported
(PartialOrd is not implemeted for Size2D), so I replace it to 
`size.greater_than(Size2D::new(MAX_CANVAS_SIZE, MAX_CANVAS_SIZE)).none()`

mrobinson commented on components/script/canvas_state.rs:
```diff
@@ -60,6 +60,39 @@ use crate::dom::paintworkletglobalscope::PaintWorkletGlobalScope;
+        }
+    }
+
+    fn get_paintable_size(&self) -> Size2D<u64> {
```
Getters in Rust usually don't use `get_`: https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter:

```suggestion
    fn paintable_size(&self) -> Size2D<u64> {
```

tharkum replied:
I know about Rust naming policy, but I was confused by existed getter methods names (like get_ipc_renderer, get_canvas_id), so I decided that in Servo are using non Rust naming policy is OK.

sagudev replied:
Servo is very old and predates Rust naming policy, so I think general rule of thumb is that new stuff try to follow it, but existing stuff can keep the old names.

mrobinson commented on components/script/canvas_state.rs:
```diff
@@ -60,6 +60,39 @@ use crate::dom::paintworkletglobalscope::PaintWorkletGlobalScope;
+
+#[derive(Clone, Debug, JSTraceable, MallocSizeOf)]
+/// Helps observe states of canvas draw target.
+struct DrawTarget {
```
The use of `DrawTarget` here is a little bit confusing, because this isn't the target of the draw itself, but metadata about it. I would call it `DrawTargetSize`.

```suggestion
struct DrawTargetSize {
```

tharkum replied:
Yes, it has mostly metadata in itself, but the goal is to have mirrored version (with size + state) of canvas draw target but on script thread. Later I have plan to add delayed creation of draw target for Canvas2D context (was cutoff from current version of PR) and it will add new filed "created/allocated" to this struct.

I was inspired by existed WebGPU implementation in Servo - it also have the same thing - metadata version of WebGPU drawing buffer on script thread.
https://github.com/servo/servo/blob/main/components/script/dom/webgpu/gpucanvascontext.rs#L54

Canvas2D [output bitmap](https://html.spec.whatwg.org/multipage/canvas.html#2dcontext:output-bitmap) -> CanvasState::DrawTarget
WebGPU [drawing buffer](https://gpuweb.github.io/gpuweb/#dom-gpucanvascontext-drawingbuffer-slot) -> GPUCanvasContext:DrawingBuffer
WebGL [drawing buffer](https://registry.khronos.org/webgl/specs/latest/1.0/#2.2) -> None

Anyway if after you and/or @sagudev decide to rename it to DrawTargetSize - no problem.

mrobinson replied:
Okay. That plan sounds interesting. For now let's give it a name that matches what it does. It's no problem to rename it later.

sagudev replied:
> I was inspired by existed WebGPU implementation in Servo - it also have the same thing - metadata version of WebGPU drawing buffer on script thread.

The reason why we named it this way in WebGPU is because it matches the spec and it's also used referenced in spec steps (this way also out code matches spec better).

But here there are no exact steps written out in spec so either name is fine by me.

mrobinson replied:
Okay. Let's call this `DrawTargetSize` then. Thanks!

mrobinson commented on components/script/canvas_state.rs:
```diff
@@ -60,6 +60,39 @@ use crate::dom::paintworkletglobalscope::PaintWorkletGlobalScope;
+const MAX_CANVAS_SIZE: u64 = 65535;
+
+#[derive(Clone, Debug, JSTraceable, MallocSizeOf)]
+/// Helps observe states of canvas draw target.
```
Please put documentation above the `derive` line and perhaps expand to:

```suggestion
/// Information about the canvas draw target size and whether the size is valid.
```

mrobinson commented on components/script/canvas_state.rs:
```diff
@@ -1892,6 +1951,19 @@ impl CanvasState {
     }
 }
 
+pub(crate) fn is_supported_canvas_size(size: Size2D<u64>) -> bool {
```
Does this need to be `pub(crate)`? I would actually make this a static method on `DrawTargetSize` above.

```rust
impl DrawTargetSize
 ...
 fn is_supported_canvas_size()
...
```

Then you can call it using `Self::is_supported_canvas_size(...)`.

tharkum replied:
Moved to DrawTarget struct. However may be better to move it to https://github.com/servo/servo/blob/main/components/shared/canvas/lib.rs  ??

mrobinson replied:
I would keep this in the `impl` block of `DrawTargetSize`.

mrobinson reviewed (changes requested):
Looks good, but I have a few very small suggestions:

sagudev commented on components/script/canvas_state.rs:
```diff
@@ -60,6 +60,46 @@ use crate::dom::paintworkletglobalscope::PaintWorkletGlobalScope;
+        }
+    }
+
+    fn paintable_size(&self) -> Size2D<u64> {
+        if self.paintable {
+            self.size
+        } else {
+            Size2D::zero()
+        }
+    }
```
So we will use paintable size for toDataUrl,toBlobUrl and similar (hence in edge cases we will error), but when rendering we will simply return empty htmlcanvassource (transparent black image as handled by WR/Layout), right? Can we document this somewhere in code.

tharkum commented:
@mrobinson @sagudev < After your suggestions I reviewed PR code and decide to simplify a little bit all this mess "DrawTarget/DrawTargetSize" thing a lot. So i simply replaced it with just "size" field and use it as "paintable" marker for canvas state.

@sagudev < Regarding "toDataUrl,toBlobUrl" - good catch!
I didn't checked it well before and now I added state validation to prevent any pixel reading IPC call for non-paintable canvas (it be handled properly on top level in different places as transparent black pixel data with natural canvas size!).
After your "snapshot" PR will be landed it will be much more better handled in one place as `snapshot::cleared()`.

sagudev commented on components/script/dom/canvasrenderingcontext2d.rs:
```diff
@@ -151,7 +141,14 @@ impl CanvasContext for CanvasRenderingContext2D {
 
     fn get_image_data(&self) -> Option<Vec<u8>> {
-        Some(self.get_rect(Rect::from_size(self.size().cast())))
+        if self.canvas_state.is_paintable() {
+            Some(
+                self.canvas_state
+                    .get_rect(self.size(), Rect::from_size(self.size())),
+            )
+        } else {
+            None
+        }
```
Shouldn't we do this as part of get_rect?

tharkum replied:
Fixed as `CanvasState::get_rect() -> Option<Vec<u8>>`

mrobinson reviewed (approved):
Looks good. Sorry for the long delay in reviewing!

tharkum commented:
@mrobinson @sagudev  < Please review.

I rebased it on top of "main" branch with "snapshot" feature.

TimvdLippe commented:
FYI the mentioned issues haven't been closed as they weren't part of separate lines. GitHub only closes the first issue in such a list.

sagudev commented:
> FYI the mentioned issues haven't been closed as they weren't part of separate lines. GitHub only closes the first issue in such a list.

Done.
